### ISSUE TITLE
How much the loss should be, if the training runs good?

### ISSUE THREAD

**Issue by HuangBo-Terraloupe** [yhenon/keras-frcnn#112]
I train the network on my dataset, not VOC, and the total loss stop decrease at around 0.8.
And 4 losses are some value like:  rpn_cls: 0.12 rpn_regr: 0.30 detector_cls: 0.30 detector_regr: 0.13
Is the value of loss are good? or I need more training data to continue decrease the losses.

right now I totally have 17k samples,
and I train it already 200 epochs.

**Comment by HuangBo-Terraloupe** [comment]
![4](https://user-images.githubusercontent.com/26878250/28328673-8ee3c6e4-6be7-11e7-90de-0f9140161f4c.png)

The performance is not good enough. Especially I got small object need to detect.

**Comment by ftiguidou** [comment]
Did you try to play with the anchor size for detecting smaller objects ?

**Comment by HuangBo-Terraloupe** [comment]
No, you mean just use small anchor size?

**Comment by ftiguidou** [comment]
I believe using smaller anchor allows you to detect smaller objects, or at least more accurately

**Comment by shuuchen** [comment]
Hi, I am also training new classes.
Is it necessary to modify the source code to add the new classes?

**Comment by HuangBo-Terraloupe** [comment]
@shuuchen you need modify the source code to add new classes.

**Comment by HuangBo-Terraloupe** [comment]
@ftiguidou originally there ar 9 different anchors, you mean only use smallest one? 
Did someone use different optimizer, maybe adam?

**Comment by shuuchen** [comment]
Thanks!
Where should I add the new classes in the code?
I found that the numbers and names of all classes are input from the .xml files under annotation directory. Is it enough to just add new classes in the .xml files?

**Comment by josifovski** [comment]
@HuangBo-Terraloupe  Here are some indications of the losses when the training is done well in the specific case the author mentioned in [this issue](https://github.com/yhenon/keras-frcnn/issues/24#issuecomment-298036580). I get similar values when training on my own dataset. Some more information on losses if the model is overfitting you can find [here](https://github.com/yhenon/keras-frcnn/issues/54).

**Comment by josifovski** [comment]
@shuuchen  If you train on a new dataset, the classes can be inferred from your annotations file, no need to modify the code. In the [readme file](https://github.com/yhenon/keras-frcnn/blob/master/README.md) it is described how to generate this file, you do not need to create XML annotations as for the VOC dataset.

**Comment by ftiguidou** [comment]
@HuangBo-Terraloupe , Not necessary use the smallest one, you can just change the values. I think remember the smallest is 128 or 64, maybe you can divide by 2 again ?

**Comment by thealchemist89** [comment]
@shuuchen you may actually use "labelimg" tool to help you annotate images with the class names that you want. It outputs in pascal voc format compatible for use in this source repository.

**Comment by shuuchen** [comment]
@thealchemist89 
@josifovski 
@HuangBo-Terraloupe 
Thanks! I tried to label some images using labelimg software.

**Comment by thealchemist89** [comment]
@ftiguidou, is there a good rule of thumb on how we could adjust the scale and ratio of the anchor_box for picking up very small objects?

**Comment by shuuchen** [comment]
@thealchemist89 
The anchor scales used are 128, 256 and 512, which are too big for small objects.
You can resize the scales according to the object sizes in your labeled images.

**Comment by ftiguidou** [comment]
@thealchemist89, I don't know if there's a rule. Trying is the best way to know. If using smaller anchor size (or ratio) gives you a better result, you can keep adjusting this parameter.

**Comment by HuangBo-Terraloupe** [comment]
Hello @ftiguidou 

I have few questions, 
1, my image patches are 500*500, should I change the image size in config.py from 600 to 500?
2, For the image channel mean, should we calculate it with our own dataset instead of directly using the mean of VOC dataset?
3, self.use_horizontal_flips = True  self.use_vertical_flips = True these two operation is flip the image and create more training data, right? It does not help my to decrease the losses,  still the same number rpn_cls: 0.12 rpn_regr: 0.30 detector_cls: 0.30 detector_regr: 0.13. Do you have any idea?

anyway, I will use small anchor size and rerun the training. Thank you.

**Comment by ftiguidou** [comment]
I'm kind of new to FRCNN too, so anything I tell you is what I understood reading the original paper and working with this code. Maybe @yhenon can correct me if I'm wrong ?

1. The code is used for resizing any image to have the smallest side at 600, and keeping the ratio of the original picture. You can change it to 500, I think this part of the code is for reducing complexity so you can reduce it more in your case.

2. I thought the same at the beginning, these means are calculated with the VOC dataset so it should not be the same for other datasets. But I'm working with my own traffic lights dataset and these means work great.

3. It seems you're working with satellite pictures ? I don't know how many samples you use but horizontal or vertical flips doesn't give you a lot more information. If your dataset is already big enough, I believe it's normal that this data augmentation is not decreasing your losses. If your dataset is small, this result is weird in my opinion.

Hope it helped you !

**Comment by shuuchen** [comment]
@HuangBo-Terraloupe 
Hi, I find that your labeled image is similar to the purpose of my project.
Would you please tell me which dataset you used for training?

**Comment by HuangBo-Terraloupe** [comment]
@shuuchen 
Sorry, the dataset belongs to my company, I can not give you.

**Comment by shuuchen** [comment]
@HuangBo-Terraloupe 
OK!

**Comment by astronstar** [comment]
how much time did it cost for 200 epochs with 17k samples。and what gpus type?

**Comment by groot-1313** [comment]
Where to edit the anchor box sizes? Is it in the config.pickle file? I tried changing it there, but the moment the training begins it is reset to 128, 256 and 512.
